role_manager_b_test.go

```diff
@@ -140,8 +143,10 @@ func BenchmarkBuildRoleLinksWithDomainPatternLarge(b *testing.B) {
 
 func BenchmarkBuildRoleLinksWithPatternAndDomainPatternLarge(b *testing.B) {
 	e, _ := NewEnforcer("examples/performance/rbac_with_pattern_large_scale_model.conf", "examples/performance/rbac_with_pattern_large_scale_policy.csv")
-	e.AddNamedMatchingFunc("g", "", util.KeyMatch4)
-	e.AddNamedDomainMatchingFunc("g", "", util.KeyMatch4)
+	roleMatcher := defaultrolemanager.NewPatternMatcher(util.IsKeyMatch4Pattern, util.KeyMatch4)
+	domainMatcher := defaultrolemanager.NewPatternMatcher(util.IsKeyMatch4Pattern, util.KeyMatch4)
+	e.SetNamedRoleMatcher("g", roleMatcher)
```

Comment:
Can we merge these two steps into one API? Like ``e.AddNamedRole/DomainMatcher(util.IsKeyMatch4Pattern, util.KeyMatch4)``？

Reply:
For the accuracy of the function name, I think the two steps are clearer.

Reply:
I just think a simple API like AddMatchingFunc should be provided, and users might not have to explicitly call NewPatternMatcher. But these two steps are ok. We should provide the default PatternMatcher when creating RoleManger. Because I have seen too many users use pattern directly without adding Match Func. Even if it is stated in the document

Reply:
> Can we merge these two steps into one API?

We could provide some general instances of `PatternMatcher`.
Then the user can use one of the predefined `PatternMatcher` instances or create one himself.

> We should provide the default PatternMatcher when creating RoleManger. Because I have seen too many users use pattern directly without adding Match Func. Even if it is stated in the document

If no `PatternMatcher` is set the performance should be slightly better, because `isPattern` never gets called.
Besides, I think it is not clear, what we should use as default `PatternMatcher`.

Reply:
> If no `PatternMatcher` is set the performance should be slightly better, because `isPattern` never gets called. Besides, I think it is not clear, what we should use as default `PatternMatcher`.

I think default `PatternMatcher` only needs to satisfy the most basic  pattern functions. Such as ``KeyMatch4`` in the above example. But that doesn't matter, what matters is that it really shouldn't provide a default if it's more efficient without `PatternMatcher`.
So just follow your PR exactly without any changes!